File: backend/src/deaths_door/routes/characters.py
"""Routes for managing character roles in the game."""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from ..character import CharacterOut
from ..events import RoleIncluded, RoleRemoved, RolesIncluded
from ..game_manager import game_manager


class AddRoleResponse(BaseModel):
    """Response after adding a role to the game."""

    status: str = Field(..., description="Operation status", examples=["success"])
    included_roles: list[str] = Field(
        ...,
        description="Names of all roles now included in the game",
        examples=[["Imp", "Chef", "Butler", "Baron"]],
    )


class AddRoleMultiResponse(BaseModel):
    """Response after adding multiple roles to the game."""

    status: str = Field(..., description="Operation status", examples=["success"])
    added_count: int = Field(..., description="Number of roles successfully added", examples=[7])


router = APIRouter(prefix="/characters", tags=["Characters"])


@router.get("/list")
async def get_game_roles() -> list[CharacterOut]:
    """
    List roles that have been added to the current game via /characters/add/multi.

    This returns the active role pool that players will be randomly assigned from.
    To see ALL possible roles in the script, use GET /game/script/roles instead.
    """
    state = await game_manager.get_state()
    return [c.to_out() for c in state.get_included_roles()]


class AddRoleRequest(BaseModel):
    """Request to add a role to the game."""

    name: str = Field(
        ...,
        description="Name of the character/role to add to the game",
        examples=["Imp", "Chef", "Empath", "Baron"],
    )


@router.post("/add", responses={400: {"description": "Role not found or already included"}})
async def add_role(req: AddRoleRequest) -> AddRoleResponse:
    """Add a role to the current game's included roles."""
    state = await game_manager.get_state()
    script = state.get_script()
    if script.get_character(req.name) is None:
        raise HTTPException(status_code=400, detail=f"Role not found: {req.name}")

    new_state = await game_manager.dispatch(RoleIncluded(name=req.name))
    return AddRoleResponse(
        status="success",
        included_roles=list(new_state.included_role_names),
    )


class AddRoleMultiRequest(BaseModel):
    """Request to add multiple roles to the game."""

    names: list[str] = Field(
        ...,
        description="List of character names to add to the game",
        examples=[["Imp", "Chef", "Butler", "Baron", "Librarian", "Empath", "Mayor"]],
    )


@router.post("/add/multi", responses={400: {"description": "One or more roles not found"}})
async def add_role_multi(req: AddRoleMultiRequest) -> AddRoleMultiResponse:
    """Add multiple roles to the current game (atomic operation)."""
    state = await game_manager.get_state()
    script = state.get_script()

    # Validate all names first
    for name in req.names:
        if script.get_character(name) is None:
            raise HTTPException(status_code=400, detail=f"Role not found: {name}")

    await game_manager.dispatch(RolesIncluded(names=tuple(req.names)))
    return AddRoleMultiResponse(status="success", added_count=len(req.names))


class RemoveRoleRequest(BaseModel):
    """Request to remove a role from the game."""

    name: str = Field(
        ...,
        description="Name of the character/role to remove from the game",
        examples=["Imp"],
    )


@router.post("/remove", responses={404: {"description": "Role not found in included roles"}})
async def remove_role(req: RemoveRoleRequest) -> AddRoleResponse:
    """Remove a role from the current game's included roles."""
    state = await game_manager.get_state()
    normalized = req.name.lower().strip()
    if not any(r.lower().strip() == normalized for r in state.included_role_names):
        raise HTTPException(status_code=404, detail=f"Role not in game: {req.name}")

    new_state = await game_manager.dispatch(RoleRemoved(name=req.name))
    return AddRoleResponse(
        status="success",
        included_roles=list(new_state.included_role_names),
    )

File: backend/src/routes/characters.rs
//! Routes for managing character roles in the game.

use axum::extract::State;
use axum::Json;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;
use utoipa_axum::router::OpenApiRouter;
use utoipa_axum::routes;

use crate::app::{AppJson, AppResult, AppState};
use crate::character::CharacterOut;
use crate::events::EventPayload;

pub fn router() -> OpenApiRouter<AppState> {
    OpenApiRouter::new()
        .routes(routes!(get_game_roles))
        .routes(routes!(add_role))
        .routes(routes!(add_role_multi))
        .routes(routes!(remove_role))
}

#[derive(Serialize, ToSchema)]
pub struct AddRoleResponse {
    pub status: String,
    /// Names of all roles now in the game's pool.
    pub included_roles: Vec<String>,
}

#[derive(Serialize, ToSchema)]
pub struct AddRoleMultiResponse {
    pub status: String,
    pub added_count: usize,
}

/// List the roles currently in the game's pool.
///
/// This is the active pool players are randomly assigned from (built via
/// `/characters/add/multi`). For *all* roles defined by the script, use
/// `GET /game/script/roles` instead.
#[utoipa::path(
    get, path = "/characters/list", tag = "Characters",
    responses((status = 200, description = "Included roles", body = [CharacterOut]))
)]
async fn get_game_roles(State(state): State<AppState>) -> AppResult<Json<Vec<CharacterOut>>> {
    let game = state.manager.get_state().await?;
    Ok(Json(
        game.get_included_roles()
            .iter()
            .map(|c| c.to_out())
            .collect(),
    ))
}

#[derive(Deserialize, ToSchema)]
pub struct AddRoleRequest {
    /// Character name to add to the pool, e.g. `Imp`.
    #[schema(example = "Imp")]
    pub name: String,
}

/// Add a single role to the game's pool.
#[utoipa::path(
    post, path = "/characters/add", tag = "Characters",
    request_body = AddRoleRequest,
    responses(
        (status = 200, description = "Role added", body = AddRoleResponse),
        (status = 400, description = "Role not found in script")
    )
)]
async fn add_role(
    State(state): State<AppState>,
    AppJson(req): AppJson<AddRoleRequest>,
) -> AppResult<Json<AddRoleResponse>> {
    // Ensure a game exists; the name is validated against the script inside
    // dispatch.
    state.manager.get_state().await?;

    let new_state = state
        .manager
        .dispatch(EventPayload::RoleIncluded {
            name: req.name.clone(),
        })
        .await?;
    Ok(Json(AddRoleResponse {
        status: "success".to_string(),
        included_roles: new_state.included_role_names.clone(),
    }))
}

#[derive(Deserialize, ToSchema)]
pub struct AddRoleMultiRequest {
    /// Character names to add to the pool (validated atomically).
    #[schema(example = json!(["Imp", "Baron", "Chef", "Empath", "Mayor"]))]
    pub names: Vec<String>,
}

/// Add multiple roles to the game's pool in one atomic call.
///
/// All names are validated against the script first; if any is unknown, none
/// are added. This is the usual way to build the role pool before adding players.
#[utoipa::path(
    post, path = "/characters/add/multi", tag = "Characters",
    request_body = AddRoleMultiRequest,
    responses(
        (status = 200, description = "Roles added", body = AddRoleMultiResponse),
        (status = 400, description = "One or more roles not found in script")
    )
)]
async fn add_role_multi(
    State(state): State<AppState>,
    AppJson(req): AppJson<AddRoleMultiRequest>,
) -> AppResult<Json<AddRoleMultiResponse>> {
    // Ensure a game exists; every name is validated against the script inside
    // dispatch, so an unknown role rejects the whole batch.
    state.manager.get_state().await?;

    let added_count = req.names.len();
    state
        .manager
        .dispatch(EventPayload::RolesIncluded { names: req.names })
        .await?;
    Ok(Json(AddRoleMultiResponse {
        status: "success".to_string(),
        added_count,
    }))
}

#[derive(Deserialize, ToSchema)]
pub struct RemoveRoleRequest {
    /// Character name to remove from the pool (case-insensitive).
    #[schema(example = "Imp")]
    pub name: String,
}

/// Remove a role from the game's pool.
#[utoipa::path(
    post, path = "/characters/remove", tag = "Characters",
    request_body = RemoveRoleRequest,
    responses(
        (status = 200, description = "Role removed", body = AddRoleResponse),
        (status = 404, description = "Role not in the game's pool")
    )
)]
async fn remove_role(
    State(state): State<AppState>,
    AppJson(req): AppJson<RemoveRoleRequest>,
) -> AppResult<Json<AddRoleResponse>> {
    // Ensure a game exists; pool membership is validated inside dispatch.
    state.manager.get_state().await?;

    let new_state = state
        .manager
        .dispatch(EventPayload::RoleRemoved {
            name: req.name.clone(),
        })
        .await?;
    Ok(Json(AddRoleResponse {
        status: "success".to_string(),
        included_roles: new_state.included_role_names.clone(),
    }))
}
